### PR TITLE
Alternative producer implementation

### PR DESCRIPTION
**Pull request by erikvanoosten**
Refactoring of the producer so that it handles errors per record.

### CODE REVIEW

File: zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala [hunk 1]
```diff
@@ -457,53 +460,63 @@ private[producer] final class ProducerLive(
...
+    for {
+      done <- Promise.make[Throwable, RecordMetadata]
+      _ <- ZIO
+             .attempt[Any] {
+               p.send(
+                 record,
+                 (metadata: RecordMetadata, err: Exception) =>
+                   unsafeRun {
+                     if (err == null) done.succeed(metadata)
+                     else done.fail(err)
```
Comment (domdorn) [inline zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala:513]: if we change this to `done.fail((record,err))` we could implement a retry strategy in 494

Comment (erikvanoosten) [reply zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala:513]: I like the idea. But are you sure you want to do the retries record by record?

**Review by svroonland — APPROVED** [review]

File: zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala (hunk 1 above)
Comment (domdorn) [reply zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala:508]: I think that is the only way to not deliver messages that have been previously accepted, e.g. prevent double delivery?

Comment (erikvanoosten) [reply zio-kafka/src/main/scala/zio/kafka/producer/Producer.scala:508]: The other option is to do something after line 489. There you can `records.zip(results)` and filter those that failed. The failed records can then be retried together. The challenge is mostly to write elegant code that merges the original results with the retry results.

**Pull request merged**